### PR TITLE
Organisations

### PR DESCRIPTION
**Pull request by leohemsted**
## organisations

`organisation` is a new table that holds an organisation's branding details - its colour, a path to its logo and its display name.

services now have a nullable organisation_id. 

## branding

branding refers to how the service interprets the org's branding. it can be one of three strings (backed by a foreign key to a static table)

* `govuk` - (default) GOV.UK branding, as current
* `org` - Only use the organisation's branding
* `both` - use both the gov.uk branding and the org's branding.

## misc cleanup

the history meta wasn't ready for nullable foreign keys. now it is!

### CODE REVIEW

File: app/history_meta.py [hunk 1]
```diff
@@ -53,7 +53,12 @@ def _col_copy(col):
         col = col.copy()
         orig.info['history_copy'] = col
         col.unique = False
-        col.default = col.server_default = None
+
+        # if the column is nullable, we could end up overwriting an on-purpose null value with a default.
```
Comment (minglis) [inline app/history_meta.py]: These sort of hacks are a bad smell. the sooner we get this versioning plugin out the better.

File: app/models.py [hunk 2]
```diff
@@ -98,12 +116,21 @@ class Service(db.Model, Versioned):
         secondary=user_to_service,
         backref=db.backref('user_to_service', lazy='dynamic'))
     restricted = db.Column(db.Boolean, index=False, unique=False, nullable=False)
-    research_mode = db.Column(db.Boolean, index=False, unique=False, nullable=True, default=False)
+    research_mode = db.Column(db.Boolean, index=False, unique=False, nullable=False, default=False)
```
Comment (minglis) [inline app/models.py]: Do we know this will work?

File: migrations/versions/0046_organisations_and_branding.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    op.add_column('services', sa.Column('organisation_id', postgresql.UUID(as_uuid=True)))
+    op.add_column('services_history', sa.Column('branding', sa.String(length=255)))
+    op.add_column('services_history', sa.Column('organisation_id', postgresql.UUID(as_uuid=True)))
+
+    op.execute("INSERT INTO branding_type VALUES ('govuk'), ('org'), ('both')")
+    # insert UKVI data as initial test data. hex and crest pulled from alphagov/whitehall
+    op.execute("""INSERT INTO organisation VALUES (
+        '9d25d02d-2915-4e98-874b-974e123e8536',
+        '#9325b2',
+        'ho_crest_27px_x2.png',
```
Comment (minglis) [inline migrations/versions/0046_organisations_and_branding.py]: Where is this file? Is this a dependency between API and frontend on content?

**Comment by minglis** [comment]
Couple of minor things. 

Also do we need organisation? Noone is using and branding applies directly to service?

File: migrations/versions/0046_organisations_and_branding.py (hunk 3 above)
Comment (leohemsted) [inline migrations/versions/0046_organisations_and_branding.py]: yup - it comes from here: https://www.notifications.service.gov.uk/static/images/email-template/crests/ho_crest_27px_x2.png - file in source code is https://github.com/alphagov/notifications-admin/blob/master/app/assets/images/email-template/crests/ho_crest_27px_x2.png

Comment (minglis) [inline migrations/versions/0046_organisations_and_branding.py]: pivotal an alert on the branding so that if fails/missing/etc we know.

**Comment by leohemsted** [comment]
comments addressed, merging

**Pull request merged**